### ISSUE TITLE
Up2: HDMI: playback on hw:0,3 fails 100% of the time

### ISSUE THREAD

**Issue by plbossart** [thesofproject/linux#532]
On Up2, HDMI uses Pin6-Port0. In theory we can use hw:0,1 w/ cvt2, hw:0,2 w/ cvt3 and hw:0,3 w/ cvt4.
This works fine with hw:0,1 and hw:0,2, but fails with hw:0,3. This also fails when directly trying the third stream so *something* must be filtering HDMI3 out...

sequence to reproduce:

````
root@ubilinux4:~# amixer -Dhw:0 cset numid=28 1
numid=28,iface=MIXER,name='Pin6-Port0 Mux'
  ; type=ENUMERATED,access=rw------,values=1,items=4
  ; Item #0 'NONE'
  ; Item #1 'cvt 2'
  ; Item #2 'cvt 3'
  ; Item #3 'cvt 4'
  : values=1
root@ubilinux4:~# speaker-test -Dhw:0,1 -c2 -r48000 -t sine

speaker-test 1.1.3

Playback device is hw:0,1
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
was set period_size = 3456
was set buffer_size = 16320
 0 - Unknown
 1 - Unknown
^CTime per period = 4.298726
root@ubilinux4:~# amixer -Dhw:0 cset numid=28 2
numid=28,iface=MIXER,name='Pin6-Port0 Mux'
  ; type=ENUMERATED,access=rw------,values=1,items=4
  ; Item #0 'NONE'
  ; Item #1 'cvt 2'
  ; Item #2 'cvt 3'
  ; Item #3 'cvt 4'
  : values=2
root@ubilinux4:~# speaker-test -Dhw:0,2 -c2 -r48000 -t sine

speaker-test 1.1.3

Playback device is hw:0,2
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
was set period_size = 3456
was set buffer_size = 16320
 0 - Unknown
 1 - Unknown
^CTime per period = 3.133030
root@ubilinux4:~# amixer -Dhw:0 cset numid=28 3
numid=28,iface=MIXER,name='Pin6-Port0 Mux'
  ; type=ENUMERATED,access=rw------,values=1,items=4
  ; Item #0 'NONE'
  ; Item #1 'cvt 2'
  ; Item #2 'cvt 3'
  ; Item #3 'cvt 4'
  : values=3
root@ubilinux4:~# speaker-test -Dhw:0,3 -c2 -r48000 -t sine

speaker-test 1.1.3

Playback device is hw:0,3
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
Unable to set hw params for playback: Connection timed out
Setting of hwparams failed: Connection timed out
````

dmesg
````
[   91.798676] sof-audio sof-audio: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   92.102604] sof-audio sof-audio: error: ipc timed out for 0x60010000 size 0x6c
[   92.103078] sof-audio sof-audio: status: fw entered - code 00000005
[   92.103612] sof-audio sof-audio: error: unexpected fault 0x00000000 trace 0x00004000
[   92.104096] sof-audio sof-audio: error: waking up any trace sleepers
[   92.104544] sof-audio sof-audio: error: hw params ipc failed for stream 1
[   92.104981] sof-audio sof-audio: ASoC: sof-audio hw params failed: -110
[   92.105416]  HDMI3: ASoC: hw_params FE failed -110
[   92.105912] sof-audio sof-audio: pcm: free stream 3 dir 0
[   92.406646] sof-audio sof-audio: error: ipc timed out for 0x60030000 size 0xc
[   92.407114] sof-audio sof-audio: status: fw entered - code 00000005
[   92.407648] sof-audio sof-audio: error: unexpected fault 0x00000000 trace 0x00004000
[   92.408132] sof-audio sof-audio: error: waking up any trace sleepers
[   92.409636] sof-audio sof-audio: pcm: close stream 3 dir 0
[   93.010192] random: crng init done
[   93.010200] random: 7 urandom warning(s) missed due to ratelimiting
[   94.718691] sof-audio sof-audio: error: ipc timed out for 0x40010000 size 0x4c
[   94.719173] sof-audio sof-audio: status: fw entered - code 00000005
[   94.719706] sof-audio sof-audio: error: unexpected fault 0x00000000 trace 0x00004000
[   94.720228] sof-audio sof-audio: error: ctx_save ipc error during suspend -110
[   94.720725] sof-audio sof-audio: FW Poll Status: reg=0x1010303 successful
[   94.721676] sof-audio sof-audio: FW Poll Status: reg=0x303 successful
[   94.721686] sof-audio sof-audio: DSP core(s) enabled? 0 : core_mask 3
[   94.721717] sof-audio sof-audio: Debug PCIR: 00000010 at  00000044

````

**Comment by plbossart** [comment]
Forgot to mention that this happens with the PCM512x board

````
root@ubilinux4:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofbxtpcm512x [sof-bxt-pcm512x], device 0: Port5 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofbxtpcm512x [sof-bxt-pcm512x], device 1: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofbxtpcm512x [sof-bxt-pcm512x], device 2: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofbxtpcm512x [sof-bxt-pcm512x], device 3: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
````

**Comment by zhuyingjiang** [comment]
@plbossart The driver and the FW SHA1?

**Comment by plbossart** [comment]
kernel a5dd3186a67b ASoC: SOF: topology: remove snd_sof_free_topology
firmware 03ababfb scripts: modify sof_bootone.sh to exit if sof_remove failed

**Comment by plbossart** [comment]
modprobe sof-pci-dev
speaker-test -Dhw:0,0 -c2 -r48000 -t sine (works)
amixer -Dhw:0 cset numid=28 3
speaker-test -Dhw:0,3 -c2 -r48000 -t sine
root@ubilinux4:~# speaker-test -Dhw:0,3 -c2 -r48000 -t sine

speaker-test 1.1.3

Playback device is hw:0,3
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
Unable to set hw params for playback: Connection timed out
Setting of hwparams failed: Connection timed out

[dmesg_hdmi3_fail.log](https://github.com/thesofproject/linux/files/2774886/dmesg_hdmi3_fail.log)

**Comment by plbossart** [comment]
another way to make HDMI crap out is when the mixers are not configured. This also leads to a DSP issue.

modprobe sof-pci-dev
amixer -Dhw:0 cset numid=28 0
root@ubilinux4:~# speaker-test -Dhw:0,1 -c2 -r48000 -t sine

speaker-test 1.1.3

Playback device is hw:0,1
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
Unable to set hw params for playback: No such device
Setting of hwparams failed: No such device
[dmesg_nocvt.log](https://github.com/thesofproject/linux/files/2774902/dmesg_nocvt.log)

**Comment by ranj063** [comment]
@plbossart a dumb question about HDMI pipelines. WHy do we need 3 pipelines for HDMI?
The way I understood it on GLK, I thought each one was meant for a different physical pin/usb c port.
Why do we need 3 for Up2?

**Comment by plbossart** [comment]
@ranj063 The graphics part is a bit complicated with notions of ports/pins/converters, but in a nutshell the hardware supports 3 independent audio streams. The Up2 supports one HDMI connector and one DP1.2, so with the latter set up in DP MST (which enables daisy chaining of displays) you can indeed enable 3 streams in parallel. See https://github.com/thesofproject/linux/issues/504.

what I was trying with this issue is to make sure 3 streams can be supported, one at a time over a single connector, before we try 3 in parallel. It's the moral equivalent of trying to play on each SSP before you try to play on all of them.

Whether the use of 3 streams in parallel makes sense from a user-experience perspective is another story. All I care about is that the SOF support deals with HDMI and all combinations are supported. Product decisions are done by product owners.

**Comment by ranj063** [comment]
> @ranj063 The graphics part is a bit complicated with notions of ports/pins/converters, but in a nutshell the hardware supports 3 independent audio streams. The Up2 supports one HDMI connector and one DP1.2, so with the latter set up in DP MST (which enables daisy chaining of displays) you can indeed enable 3 streams in parallel. See #504.
> 
> what I was trying with this issue is to make sure 3 streams can be supported, one at a time over a single connector, before we try 3 in parallel. It's the moral equivalent of trying to play on each SSP before you try to play on all of them.
> 
> Whether the use of 3 streams in parallel makes sense from a user-experience perspective is another story. All I care about is that the SOF support deals with HDMI and all combinations are supported. Product decisions are done by product owners.

@plbossart thanks for educating me on this.

**Comment by xiulipan** [comment]
@plbossart 
Confirmed. 
I though we need 3 monitors to let 3 HDMI playback work. We only tried set **cvt3** for **Pin6-Port0 Mux** and test with **hw:0,2**.

Test with a HDMI monitor connect on Pin6-Port0
Pin6-Port0 Mux cvt2 and hw:0,1 and Pin6-Port0 Mux cvt3 and hw:0,2 works fine for me.
Pin6-Port0 Mux cvt4 and hw:0.3 won't work.

@keqiaozhang @keyonjie 
With more combination, such as Pin6-Port0 Mux NULL. will also cause DSP panic.
I remember that there will only be a codec warning like below
```bash
[  580.492323] HDMI HDA Codec ehdaudio0D2: ASoC: can't set intel-hdmi-hifi2 hw params: -19
[  580.492331]  iDisp2: ASoC: hw_params BE failed -19
[  580.492333]  HDMI2: ASoC: hw_params BE failed -19
```
But now there will be following an PCM_FREE IPC ??
```bash
[  580.492338] sof-audio sof-audio: pcm: free stream 2 dir 0
[  580.492354] sof-audio sof-audio: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[  580.492417] sof-audio sof-audio: error : DSP panic!
[  580.492425] sof-audio sof-audio: panic: dsp_oops_offset 788480 offset 788480
[  580.492429] sof-audio sof-audio: status: fw entered - code 00000005
[  580.492597] sof-audio sof-audio: error: runtime exception
[  580.492600] sof-audio sof-audio: error: trace point 00004000
[  580.492603] sof-audio sof-audio: error: panic happen at :0
[  580.492611] sof-audio sof-audio: error: DSP Firmware Oops
[  580.492618] sof-audio sof-audio: error: Exception Cause: LoadStorePIFDataErrorCause, Synchronous PIF data error during LoadStore access
```

I think the panic is cause by this un-reasonable PCM_FREE IPC.
But cvt4 with hw:0.3 will still have a Codec error now
```bash
[  895.575088] HDMI HDA Codec ehdaudio0D2: ASoC: can't set intel-hdmi-hifi3 hw params: -19
[  895.575094]  iDisp3: ASoC: hw_params BE failed -19
[  895.575096]  HDMI3: ASoC: hw_params BE failed -19
```

I will look into the `PCM_FREE IPC` first, then check about the config about HDMI cvt combination.

**Comment by plbossart** [comment]
Thanks @xiulipan, let's add playback on all 3 HDMI streams as a QA check, even with a single monitor.

**Comment by plbossart** [comment]
This still happens with the latest firmware+driver

modprobe sof-pci-dev 
amixer -Dhw:0 cset numid=28 3
speaker-test -Dhw:0,3 -c2 -r48000 -t sine

The sof-logger points to a buffer error
host_params() error:buffer_set_size() failed, buffer_size = 8000
before invalid DMA channels are identified.

[logger-hdmi3.log](https://github.com/thesofproject/linux/files/2829952/logger-hdmi3.log)

**Comment by wenqingfu** [comment]
fixed by thesofproject/sof#959
